RecoLocalCalo/EcalRecProducers/plugins/EcalRecHitProducerGPU.cc

```diff
@@ -0,0 +1,247 @@
+  setup.get<EcalChannelStatusRcd>().get(ChannelStatusHandle_);
+
+  setup.get<EcalLaserAPDPNRatiosRcd>().get(LaserAPDPNRatiosHandle_);
+  setup.get<EcalLaserAPDPNRatiosRefRcd>().get(LaserAPDPNRatiosRefHandle_);
+  setup.get<EcalLaserAlphasRcd>().get(LaserAlphasHandle_);
+  setup.get<EcalLinearCorrectionsRcd>().get(LinearCorrectionsHandle_);
+  setup.get<EcalRecHitParametersGPURecord>().get(recHitParametersHandle_);
+
+  auto const& ADCToGeVConstantProduct = ADCToGeVConstantHandle_->getProduct(ctx.stream());
+  auto const& IntercalibConstantsProduct = IntercalibConstantsHandle_->getProduct(ctx.stream());
+  auto const& ChannelStatusProduct = ChannelStatusHandle_->getProduct(ctx.stream());
+
+  auto const& LaserAPDPNRatiosProduct = LaserAPDPNRatiosHandle_->getProduct(ctx.stream());
+  auto const& LaserAPDPNRatiosRefProduct = LaserAPDPNRatiosRefHandle_->getProduct(ctx.stream());
+  auto const& LaserAlphasProduct = LaserAlphasHandle_->getProduct(ctx.stream());
+  auto const& LinearCorrectionsProduct = LinearCorrectionsHandle_->getProduct(ctx.stream());
+  auto const& recHitParametersProduct = recHitParametersHandle_->getProduct(ctx.stream());
+
+  // set config ptrs : this is done to avoid changing things downstream
+  configParameters_.ChannelStatusToBeExcluded = recHitParametersProduct.ChannelStatusToBeExcluded;
+  configParameters_.ChannelStatusToBeExcludedSize = std::get<0>(recHitParametersHandle_->getValues()).get().size();
+  configParameters_.expanded_v_DB_reco_flags = recHitParametersProduct.expanded_v_DB_reco_flags;
+  configParameters_.expanded_Sizes_v_DB_reco_flags = recHitParametersProduct.expanded_Sizes_v_DB_reco_flags;
+  configParameters_.expanded_flagbit_v_DB_reco_flags = recHitParametersProduct.expanded_flagbit_v_DB_reco_flags;
+  configParameters_.expanded_v_DB_reco_flagsSize = std::get<3>(recHitParametersHandle_->getValues()).get().size();
+
+  // bundle up conditions
+  ecal::rechit::ConditionsProducts conditions{ADCToGeVConstantProduct,
+                                              IntercalibConstantsProduct,
+                                              ChannelStatusProduct,
+                                              //
+                                              LaserAPDPNRatiosProduct,
+                                              LaserAPDPNRatiosRefProduct,
+                                              LaserAlphasProduct,
+                                              LinearCorrectionsProduct,
+                                              //
+                                              IntercalibConstantsHandle_->getOffset()};
+
+  // dev mem
+  eventOutputDataGPU_.allocate(configParameters_, ctx.stream());
+
+  //
+  // schedule algorithms
+  //
+
+  edm::TimeValue_t event_time = event.time().value();
+
+  ecal::rechit::create_ecal_rehit(
+      inputDataGPU, eventOutputDataGPU_, conditions, configParameters_, nchannelsEB, event_time, ctx.stream());
```

Comment:
If this function indeed only queues a kernel call to the CUDA stream, this EDProducer would not have to be `edm::ExternalWork`.